[2단계 - 장기] 비타(오연택) 미션 제출합니다.

taek2222 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

`2.1`과 `2.2` 구현 과정에서 늘어난 비즈니스 로직과 필드로 객체지향 생활체조 요구 사항을 많이 지키지 못한 거 같습니다. 😢
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

> 선택 요구사항
필수가 아니다. 필수 요구사항 구현 후 선택적으로 구현한다.
장기 게임방을 만들고 장기 게임방에 입장할 수 있는 기능을 추가한다.

위 요구사항도 추가하여 구현했습니다! ☺️

## JDBC를 활용한 DB 연결 📌

이번 미션에서 `JDBC`를 통해 DB 연결을 구현하는 과정이 생소했습니다.

이전에 잠깐 `JPA`를 활용한 경험이 있지만, 충분한 이해가 없었던 상태에서 `JDBC`로 새롭게 접근하다 보니 고민과 어려움이 많았습니다. 

미션의 2.1, 2.2 단계에서 `JDBC`로 `DB`를 연결하며 새롭게 학습한 부분이 많았는데, 이 부분의 설계와 구현 방식에 문제가 없는지 봐주시면 감사하겠습니다.

- `SQL` 쿼리 작성과 코드 가독성 측면에서 개선이 필요한지

- `DB`와 비즈니스 코드의 테스트도 필요한지

---

### 상속 관계에서의 필드 접근 방식 📌

상속 관계에서 부모 클래스의 필드 값을 하위 클래스에서 사용할 때, 일반적으로 `getter`메서드를 통해 접근하는 것이 좋을지, 아니면 필드를 `protected`로 설정하여 하위 클래스 및 같은 패키지 내에서는 직접 접근할 수 있도록 허용해도 좋을지 고민이 됩니다. 🤔

---

바쁘신 와중에 리뷰해 주시는 것에 항상 진심으로 감사드립니다. 🙇🏻
리뷰를 받는 동안 계속해서 적극적으로 리팩토링하며 더욱 읽기 좋은 코드를 만들겠습니다. 🚀

혹시,, 아직 제가 부족한 부분이 많아 이번 `장기` 미션으로 더욱 성장해 보고 싶습니다.
매트가 가능하다면 월요일 이후에도 추가적으로 리뷰를 부탁드리고 싶습니다...! 😢

---

### Docker Compose 자료 📚
``` 
version: "3.9"
services:
  db:
    image: mysql:8.0.28
    platform: linux/x86_64
    restart: always
    ports:
      - "13306:3306"
    environment:
      MYSQL_ROOT_PASSWORD: root
      MYSQL_DATABASE: janggi
      MYSQL_USER: user
      MYSQL_PASSWORD: password
      TZ: Asia/Seoul
    volumes:
      - ./db/mysql/data:/var/lib/mysql
      - ./db/mysql/config:/etc/mysql/conf.d
      - ./db/mysql/init:/docker-entrypoint-initdb.d
```

### SQL 자료 📚
```
CREATE TABLE game (
    name VARCHAR(100) PRIMARY KEY,
    status ENUM('PLAYING', 'END') NOT NULL DEFAULT 'PLAYING'
);
```
---
```
CREATE TABLE piece (
    id INT AUTO_INCREMENT PRIMARY KEY,
    game_name VARCHAR(100) NOT NULL,
    player_team ENUM('HAN', 'CHO') NOT NULL,
    piece_type ENUM('CANNON', 'CHARIOT', 'ELEPHANT', 'GUARD', 'HORSE', 'KING', 'SOLDIER') NOT NULL,
    row INT NOT NULL,
    column INT NOT NULL,
    FOREIGN KEY (game_name, player_team) REFERENCES player(name, team) ON DELETE CASCADE
);
```
---
```
CREATE TABLE player (
    team ENUM('HAN', 'CHO') NOT NULL,
    name VARCHAR(100) NOT NULL,
    score INT DEFAULT 0,
    PRIMARY KEY (name, team),
    FOREIGN KEY (name) REFERENCES game(name) ON DELETE CASCADE
);
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **Documentation**
	- 업데이트된 README에 게임 결과 예시와 점수 표시가 추가되어, 승리, 무승부 및 게임 진행 상황을 명확하게 파악할 수 있습니다.
- **New Features**
	- 게임 진행 시 기존 게임 재개 선택, 게임 목록 조회, 게임 결과 출력 및 사용자 입력 프롬프트가 개선되어 보다 직관적인 게임 플레이 환경을 제공합니다.
	- 새로운 `KoreaChessController`가 도입되어 게임 초기화 및 로딩, 플레이 흐름을 관리합니다.
	- `Game`, `GameResult`, `GameStatus`와 같은 새로운 클래스와 열거형이 추가되어 게임 상태 및 결과를 보다 체계적으로 관리합니다.
	- 사용자 인터페이스가 개선되어 게임 선택 및 진행을 위한 새로운 메서드가 추가되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

taek2222 committed:
test: 결과값 오타 수정

taek2222 committed:
refactor: 상속 접근 제어자 정의

taek2222 committed:
feat: 예외 발생 상황도 재입력 기능 추가

taek2222 committed:
refactor: 예외 검증 메서드 분리

taek2222 committed:
refactor: 필드나 변수명에 필드명 사용 지양 개선

taek2222 committed:
feat: 길이 충분 검증 로직 추가

taek2222 committed:
test: 궁성 내부 좌표 판단 테스트 작성

taek2222 committed:
feat: 궁성 내부 판단 메서드 추가

taek2222 committed:
test: 궁성 위치 판단 경우 파라미터 추가

taek2222 committed:
feat: 궁성 좌표 판단 로직 개선

taek2222 committed:
refactor: 가독성 향상을 위한 파라미터 네이밍 변경

taek2222 committed:
test: 왕은 궁성 외부로 이동한 경우 예외 테스트 작성

taek2222 committed:
feat: 왕은 궁성 외부로 이동한 경우 예외 기능 추가

taek2222 committed:
test: 사 궁성 외부로 이동한 경우 예외 테스트 작성

taek2222 committed:
feat: 사는 궁성 외부로 이동한 경우 예외 기능 추가

taek2222 committed:
test: 가독성 줄내림 추가

taek2222 committed:
test: 기물 위치가 궁성 내부인 경우 대각선 이동 테스트 작성

taek2222 committed:
refactor: 이동의 반복 여부를 외부로 이동 및 궁성 내부 이동 판단 추가

taek2222 committed:
test: 궁성에 위치한 기물이 대각선 이동 경로 반환 테스트 작성

taek2222 committed:
test: 궁성에 위치한 기물이 대각선 이동 경로 반환 테스트 작성

taek2222 committed:
feat: 궁성에 위치한 기물이 대각선 이동 경로 반환 기능 추가

taek2222 committed:
refactor: 궁성 위치 변경 로직 삭제

taek2222 committed:
refactor: 경로 반환 로직 예외 처리 삭제

taek2222 committed:
test: 상의 이동 경로 검증 테스트 작성

taek2222 committed:
feat: 상의 이동 경로 반환 로직 검증 추가

taek2222 committed:
test: 마의 이동 경로 검증 테스트 작성

taek2222 committed:
feat: 마의 이동 경로 반환 로직 검증 추가

taek2222 committed:
test: 테스트 네이밍 일관성

taek2222 committed:
test: 궁성 내부 대각선 이동인 경우의 이동 및 예외 테스트 작성

taek2222 committed:
feat: 궁성 내부 대각선 이동인 경우의 이동 및 예외 기능 추가

taek2222 committed:
test: 잘못된 검증 과정 테스트 수정

taek2222 committed:
test: 사 궁성 내부 대각선 이동인 경우의 이동 및 예외 테스트 작성

taek2222 committed:
feat: 사 궁성 내부 대각선 이동인 경우의 이동 및 예외 기능 추가

taek2222 committed:
test: 쫄병 궁성 내부 대각선 이동인 경우의 이동 및 예외 테스트 작성

taek2222 committed:
feat: 쫄병 궁성 내부 대각선 이동인 경우의 이동 및 예외 기능 추가

taek2222 committed:
test: 차 궁성 내부 대각선 이동인 경우의 이동 및 예외 테스트 작성

taek2222 committed:
feat: 차 궁성 내부 대각선 이동인 경우의 이동 및 예외 기능 추가

taek2222 committed:
test: 포 궁성 내부 대각선 이동인 경우의 이동 및 예외 테스트 작성

taek2222 committed:
feat: 포 궁성 내부 대각선 이동인 경우의 이동 및 예외 기능 추가

taek2222 committed:
refactor: 움직임 경로 반환 책임을 각 기물로 이동

taek2222 committed:
test: 삭제한 기물 카테고리 반환 추가

taek2222 committed:
feat: 삭제한 기물 카테고리 반환 기능 추가

taek2222 committed:
test: 플레이어 점수 증가 테스트 작성

taek2222 committed:
feat: 플레이어 점수 증가 메서드 기능 추가

taek2222 committed:
test: 상대 기물 잡은 후 플레이어 점수 증가 테스트 작성

taek2222 committed:
feat: 상대 기물 잡은 후 점수 증가 기능 추가

taek2222 committed:
docs: 점수 출력 기능 추가

taek2222 committed:
feat: 점수 출력 기능 추가

taek2222 committed:
docs: 종료 턴 기능 추가 및 게임 결과 무승부 출력 포멧 변경

taek2222 committed:
test: 게임 결과 테스트 코드 작성

taek2222 committed:
feat: 게임 결과 객체 생성 및 기능 구현

taek2222 committed:
feat: 추가된 게임 결과 반환 비즈니스 로직 연결

taek2222 committed:
chore: mysql 연결 의존성 추가

taek2222 committed:
feat: DB 초기 기물, 플레이어, 게임 저장 추가

taek2222 committed:
feat: 게임 진행 중 및 새로운 게임 생성 기능 추가

taek2222 committed:
feat: 게임 중간의 상황 불러오는 기능 추가

taek2222 committed:
refactor: 게임 생성 및 출력 부분 정리

taek2222 committed:
feat: 게임 업데이트 및 삭제 되는 로직 추가

taek2222 committed:
fix: 잘못된 출력 수정

taek2222 committed:
test: 테스트 실패 수정

hyeonic commented on src/main/java/repository/GameRepository.java:
```diff
@@ -0,0 +1,12 @@
+import domain.Game;
+import java.util.List;
+
+public interface GameRepository {
+    void save(final Game game);
+
+    boolean hasPlayingGame();
+
+    List<String> findGameNameAll();
+}
```
비타는 어떠한 목적으로 repository라는 키워드를 선택했으며 interface를 통한 추상화를 하였을까요?

taek2222 replied:
이전에 연습한 프로젝트 경험으로 도메인 객체를 추가,삭제, 저장하는 과정을 하나의 추상화된 저장소로 다루고 싶어 `Repository`라는 키워드를 선택했습니다. ✅

또한, 해당 저장소가 수행하는 동작들은 `Interface`로 분리해두었는데, 이는 구체적인 구현과 상관없이 오직 **`행위`** 에만 의존할 수 있도록 하기 위함입니다. 이렇게 추상화를 통해 구현체가 변경되더라도 클라이언트 코드에는 영향을 주지 않도록 설계했습니다. 📝

hyeonic replied:
추상화를 고려하는 것은 좋지만 현재 구조에서는 큰 이점을 잘 모르겠습니다. 비타는 `Dao`와 `Repository`에 대한 차이를 알고 있나요?

taek2222 replied:
좋은 피드백 감사합니다!

이전 `Spring`으로 프로젝트를 진행할 때 개념에 대한 이해 없이 `Repository`를 바로 `DB` 연결 용도로 사용하는 습관이 있었던 것 같습니다... 😢

말씀 주신 것처럼, 현재 구조는 도메인 규칙이나 복잡한 쿼리 조합 없이 단순히 `DB` 접근 역할만을 수행하고 있어서, `Repository`보다는 `DAO`에 가까운 형태가 맞다고 판단해 네이밍을 변경하였습니다.

추가로 학습을 이어가면서 `DAO`는 일반적으로 테이블과 1:1로 매핑되는 정적인 구조를 가지는 것으로 이해했습니다. 

하지만, 그런 부분까지도 더 잘 반영하고자 리팩토링을 고민해보았지만, 이번에는 시간적 이유로 적용하지는 못했습니다. 방학 동안 충분히 고민하고, 구조 개선까지 이어가 보려고 합니다. 🤔

**변경 코멘트 :**
- [refactor: 패키지 및 클래스 네이밍 dao로 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/a18794bc67fbd28c399696309ad801cb847f8907)

hyeonic commented on src/main/java/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
```
printStackTrace을 그대로 노출하여 사용했을 때 어떠한 단점이 있을까요?

taek2222 replied:
`printStackTrace` 사용에 짚어주셔서 감사합니다. 새로운 지식을 알아갈 수 있는 계기가 되었습니다. 🙇🏻

### 문제 추적의 어려움 ⚠️
- 에러를 출력할 경우 `System.err`에 출력으로 파일이나 데이터베이스에 리다이렉션하기 어려워 로그 파일에 남지 않고, 로깅 레벨도 없어 문제 추적의 여려움이 존재합니다.

### 과도한 출력 자료 ⚠️
`printStackTrace` 에러를 출력할 경우 에러의 상세한 정보가 표현됩니다. 다만, 출력량이 많아 내부에 문자열 객체를 생성하는 과정에서 다소 시간이 소요되고, `GC` 과정에서도 메모리 사용량이 증가해 서버의 부담감이 커질 수 있습니다.

### Reflection ⚠️
에러 출력 과정에서 여러 `Synchronized`, `Reflection` 과정으로 오버헤드가 발생할 수 있습니다.

[참고 자료](https://seantech.tistory.com/216)

---

혹시, 추가적으로 궁금한 점은 의존성을 추가하면 가능하지만, 현재 상황에서는 `e.printStackTrace()` 사용하지 않고, `e.getMessage()` 사용이 가장 최선인가요? 🤔

하지만 이렇게 하면 예외의 상세한 `스택트레이스`를 확인할 수 없어, 문제 원인을 파악하기 어려운 점이 있어서 고민입니다. 혹시 이런 상황에서 더 나은 대안이 있을까요? 🤔

---

**변경 커밋:**
- [fix: 예외 처리 스택트레이스 출력 삭제](https://github.com/woowacourse/java-janggi/pull/116/commits/dfe039337d1d3a05495c7dfade14ddc0dbd5fbe2)

hyeonic replied:
먼저 e.printStackTrace()를 활용하는 방법은 말그대로 해당 애플리케이션을 실행하는 서버 내에 표준 입출력 장치를 통해 스택 트레이스 정보를 남기는 것을 의미합니다. 즉 개발 중에는 빠르게 에러 정보를 습득할 수 있으므로 큰 문제는 없다고 생각해요. 다만 실제 장기라는 애플리케이션을 별도 서버에 올려 운행하는 경우 이러한 에러 및 로깅 메시지는 레벨을 구분하기 어렵고 표준 입출력 장치에만 의존하기 때문에 확인하기 위해서는 서버에 접근하여 로그를 직접 확인해야 하는 단점이 존재합니다.

장기를 실제 운영하기 위한 목적은 아니며 레벨 1에서 다룰 내용은 아니지만 이러한 방법을 왜 지양해야 하는지 정도는 알아보면 좋지 않을까 하여 간단히 남겨보았습니다 ㅎㅎ

hyeonic commented on src/main/java/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+    public List<String> findGameNameAll() {
+        final String query = "SELECT name FROM game";
+        try (final Connection connection = getConnection();
+             final var preparedStatement = connection.prepareStatement(query);
```
connection이 null인 경우 동작을 보장할 수 있을까요?

taek2222 replied:
개선 과정에서 다른 코멘트까지 함께 해결해보았습니다. ☺️

먼저, `null` 반환과 관련된 부분은 `DB` 연결 자체에 문제가 발생했을 경우, 정상적인 비즈니스 로직을 수행할 수 없기 때문에, `null`을 반환하는 대신 **`언체크 예외`** 를 던지는 방식으로 변경했습니다. 이로 `null 가능성` 메시지도 자연스럽게 제거할 수 있었습니다. ☺️

---

그 후, DB 연결 로직의 반복적인 구현을 줄이기 위해 여러 방향성 고민을 했었습니다. 🤔

처음에는 **`정적 유틸 메서드`** 를 통해 DB 연결을 관리하려고 했습니다. 하지만 `커넥션 풀`이 없는 상황에서 DB 연결을 정적으로 관리하고 하나의 연결 객체를 바라보는 구조가 오히려 어색하게 느껴졌습니다. 🧐

또한, 테스트 측면에서도 실제 DB에 의존하지 않고 테스트 대체 구현을 쉽게 제공할 수 있다는 장점이 있어, 인터페이스 기반으로 추상화하는 방향을 선택했습니다. ✅

---

**함께 해결한 코멘트:**
- https://github.com/woowacourse/java-janggi/pull/116#discussion_r2020173554

**변경 커밋:**
- [refactor: 장기 데이터 베이스 연결 분리](https://github.com/woowacourse/java-janggi/pull/116/commits/925412c71dc976b72619f7498f1b9b831c559d3f)

hyeonic commented on src/main/java/service/GameService.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+import repository.GameRepository;
+
+public class GameService {
+
+    private final GameRepository gameRepository;
+
+    public GameService(final GameRepository gameRepository) {
+        this.gameRepository = gameRepository;
+    }
+
+    public boolean hasPlayingGame() {
+        return gameRepository.hasPlayingGame();
+    }
+
+    public List<String> findGameNameAll() {
+        return gameRepository.findGameNameAll();
+    }
+}
```
단순히 gameRepository 메서드를 by pass하는 것으로 보이는데요, 비타가 생각하는 service는 무엇이며 어떨 때 고려하나요??

taek2222 replied:
`KoreaChess` 클래스는 `Controller` 역할을 담당하는 클래스로 인식하고, 사용자의 요청 처리와 결과 출력만 담당하도록 설계했습니다. 👋

`GameRepository`는 DB와의 통신을 책임지고 있기 때문에, `Controller`와 직접적으로 연결되는 구조는 의미적으로 어색하다고 판단했습니다.
그래서 비록 `Service` 레이어가 단순히 메서드를 전달하는 느낌이 들 수도 있지만, 각 계층의 책임을 명확히 분리하는 것이 더 유지보수에 유리하다고 판단하여 구조를 나누었습니다.

혹시 이 과정에서 제가 잘못 이해한 부분이 있거나, 더 나은 방향이 있다면 조언 주시면 감사하겠습니다! 🙇🏻

hyeonic replied:
의도는 이해하였습니다! 하지만 지금 비타가 작성하신 키워드(repository, service 등)들을 보았을 때 `레이어드 아키텍처`를 고려한 것으로 보여요. 

비타는 레이어드 아키텍쳐가 왜 등장했으며 각각의 키워드가 어떠한 역할을 하는지 알고 있나요? 추가로 장기 애플리케이션이 해당 아키텍처를 적용하기에 적합한 요구사항을 가지고 있나요?

taek2222 replied:
https://github.com/woowacourse/java-janggi/pull/116#discussion_r2028359624

위 코멘트 내용과 동일하게, `Spring` 프로젝트를 진행하면서 자연스럽게 익힌 구조를 사용했지만, 그 과정에서 각 레이어가 어떤 역할을 하는지 구체적으로 인식하지는 못했던 것 같습니다. 😢

---

### Service 도입한 이유 💡

`Controller`에서 `DB` 접근과 함께 비즈니스 로직까지 처리하는 구조가 어색하게 느껴졌습니다.
그래서 비즈니스 로직을 분리하기 위한 목적으로 `Service` 레이어를 도입했던 것 같습니다.

---

방학 기간 동안 `레이어드 아키텍처`에 대해 충분히 학습한 뒤, 더 명확한 이해를 바탕으로 **`레벨 2`** 를 시작해보겠습니다!

hyeonic replied:
좋습니다! 이후 미션 진행할 때 레이어드 아키텍처가 왜 등장했고 어떠한 복잡한 문제를 해결하고자 하는지, 그 중 우리가 레벨 1에서 집중하고 있는 도메인 영역을 어떻게 풀어가야 할지 잘 고민해보면 좋을 것 같아요 😃

hyeonic commented on src/main/java/domain/Board.java:
```diff
@@ -1,71 +1,85 @@
 
-    private static void validatePlayerPieceCapture(final Position targetPosition, final Pieces playerPieces) {
-        if (playerPieces.existByPosition(targetPosition)) {
+    private static void validatePlayerPieceCapture(final Position target, final Pieces current) {
+        if (current.existByPosition(target)) {
             throw new IllegalArgumentException("도착 위치에 아군의 기물이 존재해 이동할 수 없습니다.");
         }
     }
```
static 메서드일 필요가 있을까요?

taek2222 replied:
메서드 분리 과정에서 잘못 삽입된 거 같습니다..! 😢

---

**변경 커밋:**
- [fix: 잘못된 정적 선언 수정](https://github.com/woowacourse/java-janggi/pull/116/commits/a0e5815e68d60103e4af10bfe3710485d8bff91e)

hyeonic commented on src/main/java/KoreaChess.java:
```diff
@@ -24,45 +27,76 @@ public KoreaChess(final OutputView outputView, final InputView inputView) {
+        printGameResult(board);
+    }
+
+    private void processTurn(final Game game, final Player player, final Board board) {
+        while (true) {
+            try {
+                Position start = parseToPosition(inputView.readMovingPiecePosition(player));
+                Position target = parseToPosition(inputView.readTargetPiecePosition());
+                Piece moved = board.moveAndCapture(player, start, target);
+                new PieceService(new PieceRepositoryImpl()).delete(game.getName(), player.getTeam(), target);
+                new PieceService(new PieceRepositoryImpl()).update(game.getName(), player.getTeam(), start, moved);
+                outputView.printBoard(board);
+                return;
+            } catch (Exception e) {
+                outputView.printErrorMessage(e.getMessage());
+            }
+        }
     }
```
어떠한 목적으로 service, repository 등 매번 turn이 시작될 때 새롭게 생성해주나요??

taek2222 replied:
필요한 시점에만 객체가 생성되어 실행되도록 했습니다..! 곰곰이 생각해보니, 매번 새로 생성해야 할 이유를 찾을 수 없었습니다. 그래서  `KoreaChessController` 내부 필드로 선언해 지속적으로 협력해 재사용할 수 있도록 개선했습니다! ☺️

---

**변경 코멘트:**
- [refactor: Service 필드 선언으로 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/3388a96ee0fa1a2a4ddcec84c3dcdba88a9b906c)

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -14,26 +16,49 @@ public Cannon(final Position position, final Directions directions) {
         super(position, directions);
     }
 
+    @Override
+    public List<Position> getPaths(final Position target) {
+        List<Position> paths = directions.getPaths(position, target);
+        if (position.isWithinPalace()) {
+            paths = PieceDirection.REPEATED_DIAGONAL.get().getPaths(position, target);
+        }
+        validatePaths(paths);
+        if (position.isWithinPalace()) {
+            validateDiagonalPaths(paths);
+        }
+        return paths;
+    }
```
궁성의 판단을 getPaths를 통해 하는 이유가 있을까요? move의 validateMove 내에서 처리될 순 없을까요??

taek2222 replied:
이 부분은 사실 설계 과정에서 가장 오랜 시간 고민했던 설계 중 하나였습니다... 😢

당시에는 더 나은 방향성을 찾지 못해, 우선 Paths 단계에서 검증을 수행하는 방식으로 구현했었습니다.

하지만 매트가 짚어준 부분이 당시 풀지 못했던 지점과 맞닿아, 이번 기회에 꼭 개선해보고 싶었습니다. 🔥

그래서, 코멘트를 반영하여 `MoveInfo`에 좌표 정보를 포함시키고, 이를 통해 궁성 여부를 판단하고 검증할 수 있도록 구조를 수정해보았습니다.

개선된 이 방식에 대해 매트의 의견이 궁금합니다! ☺️

---

**변경 코멘트:**
- [refactor: 시작 좌표도 저장하도록 변경 및 로직 개선](https://github.com/woowacourse/java-janggi/pull/116/commits/dba5c42a32377ff18ed16592baa8e4b108d9c719)
- [feat: 이동 방향 계산하는 로직 추가](https://github.com/woowacourse/java-janggi/pull/116/commits/533a65b9a6a147c8a4072540c730a99af6826931)
**... (이외 관련된 다양한 커밋)**
- [refactor: Path 과정에서 검증 제외로 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/15a96439df05444c5fe19356ece5edce4bd92481)

hyeonic commented on src/main/java/domain/Board.java:
```diff
@@ -1,71 +1,85 @@
 import java.util.Map;
 
 public record Board(
-        Map<Player, Pieces> playerPiecesMap
+        Map<Player, Pieces> gamePlayers
 ) {
```
Board 생성 시점에 players가 두명인 것을 보장할 수 있도록 적절히 예외 처리하는 것도 좋아보이네요!

taek2222 replied:
생각하지 못했던 부분이었습니다... 😮

다음에는 이런 부분까지 꼼꼼하게 챙길 수 있도록 하겠습니다!

**변경 코멘트:**
- [feat: 플레이어 인원 수 검증 기능 추가](https://github.com/woowacourse/java-janggi/pull/116/commits/bb160d09ef6c8b7792e048bc954c65d185feefe3)

hyeonic commented on src/main/java/repository/GameRepositoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    private Connection getConnection() {
+        try {
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
+            return null;
+        }
+    }
```
반복된 로직을 추상화할 순 없을까요?

taek2222 replied:
https://github.com/woowacourse/java-janggi/pull/116#discussion_r2022173775

해당 코멘트 내용과 동일하게 처리했습니다! 🙇🏻

hyeonic reviewed (changes requested):
안녕하세요 비타! 간단히 코멘트 남겼습니다. 확인해주시고 질문이 모호하다면 언제든 편하게 DM 남겨주세요~

> 바쁘신 와중에 리뷰해 주시는 것에 항상 진심으로 감사드립니다. 🙇🏻
> 리뷰를 받는 동안 계속해서 적극적으로 리팩토링하며 더욱 읽기 좋은 코드를 만들겠습니다. 🚀
> 
> 혹시,, 아직 제가 부족한 부분이 많아 이번 장기 미션으로 더욱 성장해 보고 싶습니다.
> 매트가 가능하다면 월요일 이후에도 추가적으로 리뷰를 부탁드리고 싶습니다...! 😢

물론 입니다 ㅎㅎ 충분히 이런저런 시도 해보시고 리뷰가 필요하다면 편하게 요청 주셔도 괜찮습니다.

taek2222 committed:
fix: 예외 처리 스택트레이스 출력 삭제

taek2222 committed:
refactor: 장기 데이터 베이스 연결 분리

taek2222 committed:
fix: 잘못된 정적 선언 수정

taek2222 committed:
refactor: Controller 네이밍 변경

taek2222 committed:
refactor: Service 필드 선언으로 변경

taek2222 committed:
refactor: 사용하지 않는 필드 삭제

taek2222 committed:
refactor: 시작 좌표도 저장하도록 변경 및 로직 개선

taek2222 committed:
refactor: 메서드 네이밍 변경

taek2222 committed:
test: 이동 방향을 반환하는 기능 테스트 작성

taek2222 committed:
feat: 이동 방향 계산하는 로직 추가

taek2222 committed:
test: 파라미터 테스트 추가

taek2222 committed:
test: 대각선 이동인지 판단하는 테스트 작성

taek2222 committed:
feat: 대각선 이동인지 판단하는 기능 추가

taek2222 committed:
refactor: 이동 정보에 경로 좌표 추가

taek2222 committed:
refactor: 이동 정보가 2개 이상이 아닌 경우 예외 추가 및 로직 변경

- 시작과 도착 좌표가 포함이 필수적으로 이루어지기 때문에 2개 이상이 필수이므로 검증을 추가했습니다.

- 다른 메서드의 역할에서 시작 좌표가 포함되어 일부분 추출하는 과정을 추가했습니다.

taek2222 committed:
test: 이동 정보에 좌표 필드와 검증 추가로 테스트 개선

taek2222 committed:
test: 방향을 반환하는 테스트 작성

taek2222 committed:
feat: 방향을 반환하는 기능 추가

taek2222 committed:
test: 벡터 반환 파라미터 테스트 추가

taek2222 committed:
test: 대각선 경로인지 확인하는 테스트 작성

taek2222 committed:
feat: 대각선 경로인지 확인하는 기능 추가

taek2222 committed:
test: 마지막 좌표가 궁성인지 확인 테스트 작성

taek2222 committed:
feat: 마지막 좌표가 궁성인지 확인 기능 추가

taek2222 committed:
refactor: Path 과정에서 검증 제외로 변경

taek2222 committed:
test: 검증 과정 변경으로 인한 테스트 수정

taek2222 committed:
test: 플레이어 인원 수 테스트 작성

taek2222 committed:
feat: 플레이어 인원 수 검증 기능 추가

taek2222 commented on src/main/java/db/JanggiConnectionProvider.java:
장기 DB 연결 로직을 추상화하기 위해 `JanggiConnectionProvider` 클래스를 생성해 분리했습니다.
다만, 외부 의존하는 로직을 어느 레이어에 배치하는 것이 적절할지.. 고민이 되었습니다. 🤔

현재는 임시로 `db`라는 패키지에 배치했지만,

이런 패키지 분리와 관련해서 매트만의 기준이나 방향성이 있다면 의견을 듣고 싶습니다! ☺️

hyeonic replied:
외부 의존하는 로직을 별도 분리하는 것은 좋은 것 같아요! 근데 해당 connection을 사용하는 repository와 별도로 위치하는건 어색하긴 하네요.

taek2222 replied:
외부에 두는 것이 가장 일반적이라 생각했는데, 매트의 의견을 듣고 보니 Repository 와 같은 패키지가 더욱 잘 어울리는 거 같아요! 🤔

**변경 코멘트 :**
- [refactor: 장기 연결 레이어 위치 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/cdd786d1adbe34ec0d36d7566d0ade542f053224)

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -14,26 +16,48 @@ public Cannon(final Position position, final Directions directions) {
+
     @Override
     public PieceCategory getCategory() {
         return CATEGORY;
     }
 
     @Override
-    public Cannon move(final Position position, final MoveInfos moveInfos) {
+    public Cannon move(final Position target, final MoveInfos moveInfos) {
         validateMove(moveInfos);
-        return new Cannon(position, directions);
+        return new Cannon(target, directions);
     }
 
     private void validateMove(final MoveInfos moveInfos) {
+        if (moveInfos.isDiagonalPath()) {
+            validateLastPathWithinPalace(moveInfos);
+        }
+        validateIntermediatePieceCount(moveInfos);
+        validateTargetPieceIsCannon(moveInfos);
+        validateCannonInIntermediatePath(moveInfos);
+    }
```
관련 [코멘트](https://github.com/woowacourse/java-janggi/pull/116#discussion_r2022738213) 답변 아래서 진행할게요!

궁성 관련 로직도 `validateMove()`로 이동하였네요 👍

궁, 궁성을 고려한 요구사항이 추가되어 if문을 통해 분기가 생긴 것으로 보여요. 이처럼 `이동 관련 전략` 요구사항이 추가되거나 변경될 경우 매번 각각의 기물 내에 validateMove에 로직을 직접 추가하거나 수정해야 할까요? 좀 더 유연하게 이를 대처할 순 없을까요??

taek2222 replied:
이 점에 대해 깊이 고민한 끝에, 비용 측면과 중복 제거 효율성을 고려하여 **`전략 패턴`**을 적용하기로 결정했습니다. 🚀

앞으로 이동 규칙이 확장되거나 변경되더라도, 각 전략을 독립적으로 관리하고 재사용할 수 있어 유지보수 측면에서도 유리하다고 판단했습니다.

혹시 개선할 점이나 보완할 부분이 있다면 피드백 부탁드립니다! 😊

**변경 코멘트 :**
- [refactor: 포 검증 전략 패턴 구조 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/2040c7a328a01a5c8ab2f9201bda6c871560eabd)
- [refactor: 차 검증도 전략 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/9c2dd95fa21fca028aedba5148122efc3af832ef)
... (이외 관련된 다양한 커밋)
- [refactor: 쫄/병 전략 패턴 구조 변경](https://github.com/woowacourse/java-janggi/pull/116/commits/a935e40e2a9301a8117d5e2c93a4a195837a3969)

hyeonic reviewed (changes requested):
안녕하세요 비타~  추가적으로 의견 나눠보면 좋을 것 같아 코멘트 남겨두었습니다. 확인해주시고 모호하거나 제가 놓친 질문 있다면 언제든 DM 남겨주세요!

taek2222 commented:
@coderabbitai full review

taek2222 committed:
refactor: 장기 연결 레이어 위치 변경

taek2222 committed:
refactor: 포 검증 전략 패턴 구조 변경

taek2222 committed:
style: 줄내림 적용

taek2222 committed:
refactor: 차 검증도 전략 변경

taek2222 committed:
feat: DB 연결 종료 기능 추가

taek2222 committed:
refactor: 상 검증도 전략 변경

taek2222 committed:
refactor: 사 전략 패턴 구조 변경

taek2222 committed:
refactor: 마 전략 패턴 구조 변경

taek2222 committed:
refactor: 왕 전략 패턴 구조 변경

taek2222 committed:
refactor: 쫄/병 전략 패턴 구조 변경

taek2222 committed:
test: 예외 처리 메시지 수정으로 테스트 예외 검증 변경

taek2222 committed:
refactor: 패키지 및 클래스 네이밍 dao로 변경

hyeonic commented on src/main/java/domain/piece/Piece.java:
```diff
@@ -10,29 +10,39 @@
+        }
+    }
+
+    protected void validateLastPathWithinPalace(final MoveInfos moveInfos) {
+        if (!moveInfos.isLastPathWithinPalace()) {
+            throw new IllegalArgumentException("궁성 이동의 경우 밖으로 이동할 수 없습니다.");
+        }
+    }
```
사용하지 않는 메서드는 제거해도 좋을 것 같아요 ㅎㅎ

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -2,40 +2,54 @@
         }
     }
+
+    private final List<MoveValidation> validations = List.of(
+            new DiagonalPalacePathValidator(),
+            new CannonIntermediatePieceCountValidator(),
+            new TargetPieceIsCannonValidator(),
+            new CannonInIntermediatePathValidator()
+    );
```
덕분에 반복된 if를 최소화할 수 있게 된 것 같습니다 👍

다만 java 컨벤션에 맞춰 필드는 메서드보다 상단에 위치시키는 방향도 좋아보이네요! 혹은 별도 상태를 가지고 있지 않다면 static을 고려해볼 수도 있을 것 같아요 ㅎㅎ

hyeonic commented on src/main/java/Application.java:
```diff
@@ -7,10 +17,20 @@ public static void main(String[] args) {
         OutputView outputView = new OutputView();
         InputView inputView = new InputView(new Scanner(System.in));
 
+        ConnectionProvider connectionProvider = new JanggiConnectionProvider();
+        GameDaoImpl gameRepository = new GameDaoImpl(connectionProvider);
+        PlayerDaoImpl playerRepository = new PlayerDaoImpl(connectionProvider);
+        PieceDaoImpl pieceRepository = new PieceDaoImpl(connectionProvider);
```
변경된 클래스 명에 맞춰 객체 이름도 변경되면 좋을 것 같아요!

그리고 구현체를 통해 변수를 할당하면 인터페이스를 활용한 이점을 누릴 수 없을 것 같아요. 

```suggestion
        ConnectionProvider connectionProvider = new JanggiConnectionProvider();
        GameDao gameDao = new GameDaoImpl(connectionProvider);
        PlayerDao playerDao = new PlayerDaoImpl(connectionProvider);
        PieceDao pieceDao = new PieceDaoImpl(connectionProvider);
```

왜 구현체가 아닌 추상화된 인터페이스에 의존하는 것을 지향하는지 잘 고민해보면 좋을 것 같습니다!

hyeonic commented on src/main/java/dao/GameDaoImpl.java:
```diff
@@ -0,0 +1,64 @@
+        this.connectionProvider = connectionProvider;
+    }
+
+    @Override
+    public void save(final Game game) {
+        final String query = "INSERT INTO game (name, status) VALUES (?, ?)";
+        try (final Connection connection = connectionProvider.getConnection();
+             final var preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setString(1, game.getName());
+            preparedStatement.setString(2, game.getStatus().name());
+            preparedStatement.executeUpdate();
+            connectionProvider.closeConnection(connection);
+        } catch (final SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
`try-with-resources`를 사용하면 자동적으로 connection이 close될 것을 기대하기 때문에 `connectionProvider.closeConnection(connection)`을 활용한 명시적인 close는 불필요할 것 같아요. `try-with-resources`가 어떻게 동작하는지 이번 기회를 통해 알아보면 좋을 것 같습니다!

hyeonic reviewed (approved):
안녕하세요 비타! 대부분 요구사항이 잘 반영된 것 같아 이만 머지 해도 좋을 것 같습니다 ㅎㅎ

더 논의하고 싶거나 궁금한 부분이 있다면 언제든 편하게 DM 남겨주세요.

저희가 주고 받은 대화 속에서 비타만의 근거를 만들어가는데 도움이 되었길 바라봅니다. 레벨 1 너무 고생하셨고 다음 레벨도 화이팅 👏
